cmd/controller-manager/app/config.go

```diff
@@ -17,17 +17,177 @@ limitations under the License.
+	LargeClusterSizeThreshold int32
+	UnhealthyZoneThreshold    float32
+}
+
+// PersistentVolumeBinderControllerInfo is part of context object for the controller manager.
+type PersistentVolumeBinderControllerInfo struct {
+	PVClaimBinderSyncPeriod metav1.Duration
+	VolumeConfiguration     componentconfig.VolumeConfiguration
+}
+
+// ReplicaSetInfo is part of context object for the controller manager.
+type ReplicaSetInfo struct {
+	ConcurrentRSSyncs int32
+}
+
+// ReplicationControllerInfo is part of context object for the controller manager.
+type ReplicationControllerInfo struct {
+	ConcurrentRCSyncs int32
+}
+
+// ResourceQuotaControllerInfo is part of context object for the controller manager.
+type ResourceQuotaControllerInfo struct {
+	ResourceQuotaSyncPeriod      metav1.Duration
+	ConcurrentResourceQuotaSyncs int32
+}
+
+// SAControllerInfo is part of context object for the controller manager.
+type SAControllerInfo struct {
+	ConcurrentSATokenSyncs int32
+}
+
+// ComponentConfigInfo is the main context object for the controller manager.
+type ComponentConfigInfo struct {
+	GenericComponentConfig                 GenericComponentInfo
+	DebuggingConfig                        DebuggingInfo
+	CloudProviderConfig                    CloudProviderInfo
+	AttachDetachControllerConfig           AttachDetachControllerInfo
+	CSRSigningControllerConfig             CSRSigningControllerInfo
+	DeploymentControllerConfig             DeploymentControllerInfo
+	PodGCControllerConfig                  PodGCControllerInfo
+	EndPointControllerConfig               EndPointControllerInfo
+	GarbageCollectorControllerConfig       GarbageCollectorControllerInfo
+	HPAControllerConfig                    HPAControllerInfo
+	NamespaceControllerConfig              NamespaceControllerInfo
+	NodeLifecycleControllerConfig          NodeLifecycleControllerInfo
+	PersistentVolumeBinderControllerConfig PersistentVolumeBinderControllerInfo
+	ReplicaSetConfig                       ReplicaSetInfo
+	ReplicationControllerConfig            ReplicationControllerInfo
+	ResourceQuotaControllerConfig          ResourceQuotaControllerInfo
+	SAControllerConfig                     SAControllerInfo
+	Controllers                            []string
+	ExternalCloudVolumePlugin              string
+	ConcurrentDaemonSetSyncs               int32
+	ConcurrentJobSyncs                     int32
+	NodeSyncPeriod                         metav1.Duration
+	DeletingPodsQPS                        float32
+	DeletingPodsBurst                      int32
+	RegisterRetryCount                     int32
+	ServiceCIDR                            string
+	NodeCIDRMaskSize                       int32
+	RootCAFile                             string
+}
+
 // Config is the main context object for the controller manager.
 type Config struct {
-	// TODO: split up the component config. This is not generic.
-	ComponentConfig componentconfig.KubeControllerManagerConfiguration
+	ComponentConfig ComponentConfigInfo
```

Comment:
so we got rid of the `KubeControllerManagerConfiguration` configuration? I think we have to move this great new structure to the api object.

For the controller manager we can change the component config API object easily because we have not persisted it anywhere. For the scheduler we already did that (as configuration file on disk).  So for the scheduler we have to create a new version and potentially convert it.

But he we should at least follow the component config API pattern.